### ISSUE TITLE
docker driver OOM killed detection flaky on cgroups v2

### ISSUE THREAD

**Issue by shoenig** [hashicorp/nomad#13119]
This test has started failing a lot, maybe 22.04 related 

```
=== RUN   TestDockerDriver_OOMKilled
    docker.go:36: Successfully connected to docker daemon running version 20.10.16+azure-2
    driver_test.go:2[421](https://github.com/hashicorp/nomad/runs/6595376614?check_suite_focus=true#step:4:422): not killed by OOM killer: Docker container exited with non-zero exit code: 137
--- FAIL: TestDockerDriver_OOMKilled (2.41s)
```

**Comment by tgross** [comment]
I've been able to reproduce this locally on a VM running ubuntu 22.04 (Jammy), which is configured with cgroups v2, but I've been unable to reproduce with any cgroups v1 configuration. With the following patch:

```diff
$ git diff
diff --git a/drivers/docker/handle.go b/drivers/docker/handle.go
index 0783bd901..c0f787a44 100644
--- a/drivers/docker/handle.go
+++ b/drivers/docker/handle.go
@@ -11,6 +11,7 @@ import (
        "time"

        "github.com/armon/circbuf"
+       "github.com/davecgh/go-spew/spew"
        docker "github.com/fsouza/go-dockerclient"
        "github.com/hashicorp/consul-template/signals"
        hclog "github.com/hashicorp/go-hclog"
@@ -239,6 +240,11 @@ func (h *taskHandle) run() {
        container, ierr := h.waitClient.InspectContainerWithOptions(docker.InspectContainerOptions{
                ID: h.containerID,
        })
+
+       scs := spew.NewDefaultConfig()
+       scs.DisableMethods = true
+       h.logger.Warn("container.State", "state", scs.Sdump(container.State))
+
        oom := false
        if ierr != nil {
                h.logger.Error("failed to inspect container", "error", ierr)
```

I can see the `OOMKilled` flag simply isn't being set:

```
2022-07-28T14:13:57.596Z [WARN]  docker/handle.go:246: docker: container.State: container_id=aac5d6b4620001d04314fd18bea623f5f767615778b88292b98706e4ad998263
  state=
  | (docker.State) {
  |  Status: (string) (len=6) "exited",
  |  Running: (bool) false,
  |  Paused: (bool) false,
  |  Restarting: (bool) false,
  |  OOMKilled: (bool) false,
  |  RemovalInProgress: (bool) false,
  |  Dead: (bool) false,
  |  Pid: (int) 0,
  |  ExitCode: (int) 137,
  |  Error: (string) "",
  |  StartedAt: (time.Time) {
  |   wall: (uint64) 485069344,
  |   ext: (int64) 63794614435,
  |   loc: (*time.Location)(<nil>)
  |  },
  |  FinishedAt: (time.Time) {
  |   wall: (uint64) 515940237,
  |   ext: (int64) 63794614437,
  |   loc: (*time.Location)(<nil>)
  |  },
  |  Health: (docker.Health) {
  |   Status: (string) "",
  |   FailingStreak: (int) 0,
  |   Log: ([]docker.HealthCheck) <nil>
  |  }
  | }
```

I traced this through go-dockerclient to moby/moby and ended finding this very similar test failure on the moby project: https://github.com/moby/moby/issues/41929. From there I dug down to this PR in containerd https://github.com/containerd/containerd/pull/6323 where the field they're reading from was changed. This was released in `containerd` [v1.6.3](https://github.com/containerd/containerd/releases/tag/v1.6.3) and [v1.5.12](https://github.com/containerd/containerd/releases/tag/v1.5.12). The moby project has some unusual vendoring setup going on but it looks like even the most recent tag for moby is pinned to `containerd` [v1.5.0 + security fixes](https://github.com/moby/moby/blob/v20.10.17/vendor.conf#L135).

So we can't really fix this even by updating our moby dependency until upstream has done the same. I'm going to add a flag to the test to disable it in the case of cgroups v2 with a pointer to this issue, and then mark this issue as waiting for upstream. I'll also drop a note over in https://github.com/moby/moby/issues/41929 pointing them to the containerd PR.

**Comment by tgross** [comment]
https://github.com/hashicorp/nomad/pull/13928 will remove the test flake. I've updated the labels for this issue to make it clear this is a problem with the driver and not the test.

**Comment by shoenig** [comment]
Wow nice detective work @tgross, thanks

**Comment by mr-karan** [comment]
We recently faced an issue in prod (on cgroups v2, Ubuntu 22.04), where the docker container restarted just few seconds after a restart. We think a memory build-up by the app in such a short span of time is impossible (as the normal memory usage is ~1GB, while the limits are 24GB).

Just wanted to confirm, is the issue we faced related to the above issue being discussed? 

![image](https://user-images.githubusercontent.com/5689132/231365485-59862665-719b-4669-83d5-204fbfad1486.png)

**Comment by shoenig** [comment]
Hi @mr-karan, I believe a docker container can exit with 137 whether the issue is 

- the task itself consumed memory beyond its limit
- the kernel is experiencing memory pressure and activates the cgroup oversubscription oom-killer, and chooses to kill containers
- the docker process decides the host has insufficient memory 

Usually at least one of `docker logs` or `dmesg` should be able to shed light on the underlying kill reason. 

That said, I'm currently in the process of upgrading our docker libraries which are a couple of years old, and may contain bug fixes / improvements in this area.

**Comment by alexandre-pma** [comment]
We are having a similar issue as @mr-karan in our dev environment.
We've recently updated the OS to Ubuntu 22.04 (previously on 20.04), so using cgroups v2, and our services started experiencing a lot of random OOM kills.

Our services are docker containers which run Java with `-XX:+UseContainerSupport` support by default.
Focusing on a specific service we have which is configured as:
```
      resources {
        memory = 512
        memory_max = 640
      }
```
In our staging environment (still Ubuntu 20.04), asking java about memory inside the container:
```
$ java -XshowSettings:vm -version
VM settings:
    Max. Heap Size (Estimated): 320.00M
    Using VM: Eclipse OpenJ9 VM
```
But on our dev environment (Ubuntu 22.04 host), which is using the same docker image:
```
$ java -XshowSettings:vm -version
VM settings:
    Max. Heap Size (Estimated): 4.97G
    Using VM: Eclipse OpenJ9 VM
```

So it seems that Nomad isn't able to correctly set memory on docker.
I wonder if there is any workaround on this.

**Comment by tgross** [comment]
I'm doing a bit of issue triage and realizing there are actually a number of these reports but we haven't been able to pin it down. I've collected them together in an internal issue (https://hashicorp.atlassian.net/browse/NET-12155) and I'll mark this for roadmapping the work to hunt this down.

The full collection:
* https://github.com/hashicorp/nomad/issues/13119 
* https://github.com/hashicorp/nomad/issues/9425 
* https://github.com/hashicorp/nomad/issues/17782 
* https://github.com/hashicorp/nomad/issues/23412

**Comment by EugenKon** [comment]
Another one OOM https://github.com/hashicorp/nomad/issues/23778#issuecomment-2278635279
I do not remember exact resolution, but problem was belong that nginx was not able to process received signal. Image do not use tini yet (Yes, it is very old). On that instance we have 16GB and 7GB was leased to `nginx`. For sure it has enough memory.

**Comment by EugenKon** [comment]
Another case when I see OOM, happens when I just reduce the number of jobs.
How to reproduce:
Just press "-" button and you will see OOM error message. For sure this is not OOM.
![Image](https://github.com/user-attachments/assets/a1e2f83e-f409-4e0a-84b1-b027e72aa753)

**Comment by tgross** [comment]
@EugenKon yes that's entirely understood at this point. No need to repeat information we already have. This just needs engineering time on it.

**Comment by EugenKon** [comment]
@tgross Sorry, I thought that nobody mention that this happened when "-" button is clicked.

**Comment by pkazmierczak** [comment]
This issue is no longer present in Nomad 1.9+ because we've upgraded our docker client and that in turn received patches from upstream. I'm gonna close this ticket, but feel free to re-open if you observe OOM detection flakiness in Nomad 1.9 or later.
